src/extender.cpp

```diff
@@ -0,0 +1,1088 @@
+      if (!ObjEquality()(innerPseudo->argument(), pseudo->argument())) return {};
+      return innerPseudo->selector()->elements();
+    }
+    else if (name == "has" && name == "host" && name == "host-context" && name == "slotted") {
+      // We can't expand nested selectors here, because each layer adds an
+      // additional layer of semantics. For example, `:has(:has(img))`
+      // doesn't match `<div><img></div>` but `:has(img)` does.
+      return { complex };
+    }
+
+    return {};
+
+  }
+  // EO extendPseudoComplex
+
+  // ##########################################################################
+  // Extends [pseudo] using [extensions], and returns
+  // a list of resulting pseudo selectors.
+  // ##########################################################################
+  std::vector<Pseudo_Selector_Obj> Extender::extendPseudo(
+    Pseudo_Selector_Obj pseudo,
+    ExtSelExtMap& extensions,
+    CssMediaRule_Obj mediaQueryContext)
+  {
+    auto selector = pseudo->selector();
+    SelectorListObj extended = extendList(
+      selector, extensions, mediaQueryContext);
+    if (!extended || !pseudo || !pseudo->selector()) { return {}; }
+    if (ObjEqualityFn(pseudo->selector(), extended)) { return {}; }
+
+    // For `:not()`, we usually want to get rid of any complex selectors because
+    // that will cause the selector to fail to parse on all browsers at time of
+    // writing. We can keep them if either the original selector had a complex
+    // selector, or the result of extending has only complex selectors, because
+    // either way we aren't breaking anything that isn't already broken.
+    std::vector<ComplexSelectorObj> complexes = extended->elements();
+
+    if (pseudo->normalized() == "not") {
+      if (!hasAny(pseudo->selector()->elements(), hasMoreThanOne)) {
+        if (hasAny(extended->elements(), hasExactlyOne)) {
+          complexes.clear();
+          for (auto complex : extended->elements()) {
+            if (complex->length() <= 1) {
+              complexes.push_back(complex);
+            }
+          }
+        }
+      }
+    }
+    
+    std::vector<ComplexSelectorObj> expanded = expand(
+      complexes, extendPseudoComplex, pseudo, mediaQueryContext);
+
+    // Older browsers support `:not`, but only with a single complex selector.
+    // In order to support those browsers, we break up the contents of a `:not`
+    // unless it originally contained a selector list.
+    if (pseudo->normalized() == "not") {
+      if (pseudo->selector()->length() == 1) {
+        std::vector<Pseudo_Selector_Obj> pseudos;
+        for (size_t i = 0; i < expanded.size(); i += 1) {
+          pseudos.push_back(pseudo->withSelector(
+            expanded[i]->wrapInList()
+          ));
+        }
+        return pseudos;
+      }
+    }
+
+    SelectorListObj list = SASS_MEMORY_NEW(SelectorList, "[phony]");
+    return { pseudo->withSelector(list->concat(complexes)) };
```

Comment:
Can we avoid `concat` here like this?

```c++
SelectorListObj list = SASS_MEMORY_NEW(SelectorList, "[phony]", complexes));
return { pseudo->withSelector(list); }
```

Reply:
That constructor does not exist. Since I'm tired of debating it I removed the Vectorized method chaining experiment.